### DISCUSSION TITLE
convert -sharpen slow for large images .JPG any workaround?

### DISCUSSION THREAD

**Discussion by fred888651** [ImageMagick/ImageMagick#2535] [Help]
ImageMagick-7.0.9-Q16 - Windows 10 - PowerShell 5.1
Intel i7 3,6 GHz and 16 GB ram - no other workloads in the system 

magick convert _1600x2400.jpg_ -sharpen "0x5" outfile.jpg 
**about 20 seconds for 3.2 mb jpg large file 1600x2400 pix** 

(sharpening radius/sigma is just for testing - its probably much to high)

magick convert _6400x9600.jpg_ -sharpen "0x5" outfile.jpg 
**about 420 seconds 22.7 mb jpg very large file 6400x9600 pix**
(CPU load is about 40-60% and RAM about 600-1000 MB)

**QUESTION:**
Is there some way to speed up the sharpening method (for large/very large images)? 
As I have understood -sharpen is the fastest method

I am a photographer and almost always deal with more or less large image files (with no or minimal compression)

Maybe some buffer settings?
Would I benefit from using 8 bit version of IM?

**Comment by urban-warrior** [comment]
Your 6400x9600 command ran in 11 seconds for us on a 32GB Linux system with 6 cores.  Add -debug cache to your command-line.  Chances are the 6400x9600 image is exceeding memory and the pixels are pushed to disk which of course is way slow compared to in-memory pixels.  ImageMagick can take advantage of multiple cores on your system.   4 or more cores are recommended for performance.   See https://imagemagick.org/script/architecture.php.

**Comment by snibgo** [comment]
> Would I benefit from using 8 bit version of IM?

Yes, and non-HDRI so IM works with integer pixel values.

For comparison, this command...
```
magick -size 6400x9600 xc: -sharpen 0x5 x.jpg
```
... takes 182 seconds on my i7 12GB laptop, Windows 8.1. IM doesn't need to cache to disk, and CPU is 100% throughout.

Windows "taskmgr" will show what resources are used, or look at the %TEMP% directory while the command is running.

**Comment by urban-warrior** [comment]
We have a 6 core E5-1650.  Your command completed in 47 seconds with 16-bit HDRI enabled (floating-point calculations).

**Comment by fred888651** [comment]
Thanks!

Its a 8 core i7 system with loads of free memory - disk usage is near 0% and no recent files in $env:TEMP folder 

However, I saw that Window Update was working afterwards. So maybe that did something.

I changed env variabes as follows  
**$env:MAGICK_THREAD_LIMIT = 8 
$env:MAGICK_MEMORY_LIMIT = 4gb**

magick convert IMG_0341.jpg -sharpen 0x5.0 SHARPEN_IMG_0341.jpg

about 17 seconds now - thats more acceptable

**Comment by snibgo** [comment]
Windows Update can hog resources.

Note that setting limits will set _extra_ limits. It won't relax limits (unless that mechanism has already set a stronger limit), and it won't give you more resources.

**Comment by fmw42** [comment]
If you have an appropriate GPU, you can switch to -unsharp, which uses the GPU, whereas -sharpen apparently does not, since it is not listed at https://imagemagick.org/www/script/opencl.php

**Comment by fred888651** [comment]
Thanks! Very interesting

I have a pretty decent GPU (Nvidia GTX 770) - never used OpenCL but I think that requirement is matched. Only used CUDA in 3d applications (Blender) and the speed improvement compared to GPU is dramatic. 

I do have some config files under application profile for ImageMagick dealing with my GPU so it looks like its being activated.

Good advice - thanks! 

Will try unsharp instead of blur then!

**Reply by fmw42** [reply]
Unsharp in place of sharpen not blur! I am surprised that sharpen is not on the list. Perhaps it is an oversight. Blur is already in the list of GPU.
